test/externalTests/external_test.py

```diff
@@ -0,0 +1,364 @@
+            executable='/bin/bash',
+            env=env,
+            stdout=log if not logfile else None,
+            stderr=None
+        )
+        return ret.returncode
+
+
+@dataclass
+class SolcConfig:
+    binary_type: str
+    binary_path: str
+    branch: Optional[str] = field(default="master")
+    install_dir: Optional[str] = field(default="solc")
+    solcjs_src_dir: Optional[str] = field(default="")
+
+
+@dataclass
+class TestConfig:
+    repo_url: str
+    ref_type: str
+    ref: str
+    config_file: Optional[str]
+    config_var: Optional[str]
+    build_dependency: Optional[str] = field(default="nodejs")
+    compile_only_presets: Optional[List[str]] = field(default_factory=list)
+    settings_presets: Optional[List[str]] = field(default_factory=list)
+    evm_version: Optional[str] = field(default=CURRENT_EVM_VERSION)
+    solc: Dict[str, SolcConfig] = field(
+        default_factory=lambda: defaultdict(SolcConfig))
+
+    def __post_init__(self):
+        if isinstance(self.solc, dict):
+            self.solc = SolcConfig(**self.solc)
+
+    def selected_presets(self):
+        return self.compile_only_presets + self.settings_presets
+
+
+class TestRunner(metaclass=ABCMeta):
+
+    @staticmethod
+    def on_local_test_dir(fn):
+        """Run function inside the test directory"""
+
+        def f(self, *args, **kwargs):
+            if self.test_dir:
+                os.chdir(self.test_dir)
+            else:
+                raise RuntimeError("Test directory not defined")
+            return fn(self, *args, **kwargs)
+        return f
+
+    @abstractmethod
+    def setup_environment(self, test_dir: Path):
+        pass
+
+    @abstractmethod
+    def clean(self):
+        pass
+
+    @abstractmethod
+    def compiler_settings(self, solc_version: str, presets: Tuple[str] = AVAILABLE_PRESETS):
+        pass
+
+    @abstractmethod
+    def compile(self, solc_version: str, preset: str):
+        pass
+
+    @abstractmethod
+    def run_test(self, preset: str):
+        pass
+
+
+class ExternalTest:
+    def __init__(self, config: TestConfig):
+        if config.solc.binary_type not in ('native', 'solcjs'):
+            raise RuntimeError(
+                f"Invalid solidity compiler binary type: {config.solc.binary_type}")
+        if config.solc.binary_type != "solcjs" and config.solc.solcjs_src_dir != "":
+            raise RuntimeError(
+                f"""Invalid test configuration: 'native' mode cannot be used with 'solcjs_src_dir'.
+              Please use 'binary_type: solcjs' or unset: 'solcjs_src_dir: {config.solc.solcjs_src_dir}'""")
+        self.config = config
+
+    @staticmethod
+    def parse_command_line(description: str, args: str):
+        arg_parser = ArgumentParser(description)
+        arg_parser.add_argument('--solc-binary-type', required=True, type=str,
+                                help="""Solidity compiler binary type""",
+                                choices=['native', 'solcjs'])
```

Comment:
This is our preferred style for calls in Python:
```suggestion
        arg_parser.add_argument(
            '--solc-binary-type',
            required=True,
            type=str,
            help="""Solidity compiler binary type""",
            choices=['native', 'solcjs'],
        )
```